backends/p4tools/common/core/target.cpp

```diff
@@ -48,23 +47,95 @@ bool Target::init(std::string deviceName, std::string archName) {
+
+    return instance->second->makeContext();
+}
+
+std::optional<ICompileContext *> Target::initializeTarget(std::string_view toolName,
+                                                          const std::vector<const char *> &args) {
+    // Establish a dummy compilation context so that we can use ::error to report errors while
+    // processing target and arch.
+    class DummyCompileContext : public BaseCompileContext {
+    } dummyContext;
+    AutoCompileContext autoDummyContext(&dummyContext);
+    if (args.size() < 3) {
+        ::P4::error("Missing --target and --arch arguments");
+        return std::nullopt;
+    }
+    std::optional<std::string> target;
+    std::optional<std::string> arch;
+    // Loop through arguments (skip the program name)
+    for (size_t i = 1; i < args.size(); ++i) {
+        const std::string &arg = args[i];
+        if (arg == "--arch") {
+            if (i + 1 < args.size()) {
+                arch = args[i + 1];
+            } else {
+                ::P4::error("Missing architecture name after --arch");
+                return std::nullopt;
+            }
+        }
+        if (arg == "--target") {
+            if (i + 1 < args.size()) {
+                target = args[i + 1];
+            } else {
+                ::P4::error("Missing device name after --target");
+                return std::nullopt;
+            }
+        }
+    }
+    if (!target) {
+        ::P4::error("Missing --target argument");
+        return std::nullopt;
+    }
+    if (!arch) {
+        ::P4::error("Missing --arch argument");
+        return std::nullopt;
+    }
+    return initializeTarget(toolName, target.value(), arch.value());
+}
+
+bool Target::setDevice(std::string_view deviceName) {
+    std::string lowerCaseDeviceName(deviceName);
+    std::transform(lowerCaseDeviceName.begin(), lowerCaseDeviceName.end(),
+                   lowerCaseDeviceName.begin(), ::tolower);
+    if (defaultArchByDevice.count(lowerCaseDeviceName) == 0U) {
         return false;
     }
 
-    auto archName = curTarget ? curTarget->archName : defaultArchByDevice.at(deviceName);
+    auto archName = curTarget ? curTarget->archName : defaultArchByDevice.at(deviceName.data());
     return init(deviceName, archName);
 }
 
-bool Target::setArch(std::string archName) {
-    std::transform(archName.begin(), archName.end(), archName.begin(), ::tolower);
-    if (defaultDeviceByArch.count(archName) == 0U) {
+bool Target::setArch(std::string_view archName) {
+    std::string lowerCaseArchName(archName);
+    std::transform(lowerCaseArchName.begin(), lowerCaseArchName.end(), lowerCaseArchName.begin(),
+                   ::tolower);
+    if (defaultDeviceByArch.count(lowerCaseArchName) == 0U) {
         return false;
     }
 
-    auto deviceName = curTarget ? curTarget->deviceName : defaultDeviceByArch.at(archName);
+    auto deviceName = curTarget ? curTarget->deviceName : defaultDeviceByArch.at(archName.data());
```

Comment:
```suggestion
    auto deviceName = curTarget ? curTarget->deviceName : defaultDeviceByArch.at(lowerCaseArchName);
```

Seems wrong to use the original name after asking for the lowercase one. Also, the search should fowk even for `string_view` due to use of transparent comparator in the `map`.